[World Leaks] - Investigate leaks in LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/

nipunshukla21 opened the pull request:
#### b710a47a6e5332cbbdecacd2df9aa69355e5f95b
<pre>
[World Leaks] - Investigate leaks in LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/
<a href="https://bugs.webkit.org/show_bug.cgi?id=298733">https://bugs.webkit.org/show_bug.cgi?id=298733</a>
<a href="https://rdar.apple.com/160328982">rdar://160328982</a>

Reviewed by Ryan Reno and Antti Koivisto.

This leak was found to be caused by WebCore::Style::Scope::m_anchorPositionedToAnchorMap,
which contains AnchorPositionedToAnchorEntry as its value. AnchorPositionedToAnchorEntry
is a struct which holds AnchorPositionedKey as a member, which itself holds
RefPtr&lt;const Element&gt;. Because Element holds a strong reference to Document, which holds
a strong reference to Style::Scope through Document::m_styleScope, we have a reference cycle.
This change breaks the cycle by changing the definition of AnchorPositionedToAnchorEntry so
that it no longer holds AnchorPositionedKey. This change should fix leaks in approximately
270 tests in LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/.

Test: fast/css/css-anchor-position/anchor-position-does-not-leak.html
* LayoutTests/fast/css/css-anchor-position/anchor-position-does-not-leak-expected.txt: Added.
* LayoutTests/fast/css/css-anchor-position/anchor-position-does-not-leak.html: Added.
* LayoutTests/fast/css/css-anchor-position/anchor-size-page-zoom.html:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositioningStatesAfterInterleavedLayout):
* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolve):

Canonical link: <a href="https://commits.webkit.org/300880@main">https://commits.webkit.org/300880@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1cff8ddaf552d64088950608edbc6a176ab07b15

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124134 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43827 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34544 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130961 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76244 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126011 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44568 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52427 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94425 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62645 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127088 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35510 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111038 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75017 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34459 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29201 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74448 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105250 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29422 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133635 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51063 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38910 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102897 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51443 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107257 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102705 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26139 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48057 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26307 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47941 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50919 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56693 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50376 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53726 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52052 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1cff8ddaf552d64088950608edbc6a176ab07b15. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50616

rreno reviewed (approved):
Forgot to r+ after the test was added. Thank you.

nipunshukla21 committed:
[World Leaks] - Investigate leaks in LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/
https://bugs.webkit.org/show_bug.cgi?id=298733
rdar://160328982

Reviewed by Ryan Reno and Antti Koivisto.

This leak was found to be caused by WebCore::Style::Scope::m_anchorPositionedToAnchorMap,
which contains AnchorPositionedToAnchorEntry as its value. AnchorPositionedToAnchorEntry
is a struct which holds AnchorPositionedKey as a member, which itself holds
RefPtr<const Element>. Because Element holds a strong reference to Document, which holds
a strong reference to Style::Scope through Document::m_styleScope, we have a reference cycle.
This change breaks the cycle by changing the definition of AnchorPositionedToAnchorEntry so
that it no longer holds AnchorPositionedKey. This change should fix leaks in approximately
270 tests in LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/.

Test: fast/css/css-anchor-position/anchor-position-does-not-leak.html
* LayoutTests/fast/css/css-anchor-position/anchor-position-does-not-leak-expected.txt: Added.
* LayoutTests/fast/css/css-anchor-position/anchor-position-does-not-leak.html: Added.
* LayoutTests/fast/css/css-anchor-position/anchor-size-page-zoom.html:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositioningStatesAfterInterleavedLayout):
* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolve):

Canonical link: https://commits.webkit.org/300880@main
